File: client-desktop/src/bip39.cpp
// SPDX-License-Identifier: AGPL-3.0-or-later
#include "bip39.hpp"

#include <sodium.h>     // crypto_hash_sha256

#include <QHash>
#include <QRegularExpression>
#include <QStringList>

#include <cstring>

#include "bip39_wordlist.hpp"

namespace fb::desktop {

namespace {

// Lazily-built reverse lookup: word → index. Avoids a const Map<> with a
// 2048-entry initializer; we only build it once on first phrase_to_seed.
const QHash<QString, std::uint16_t>& word_index() {
    static const QHash<QString, std::uint16_t> kIndex = []() {
        QHash<QString, std::uint16_t> m;
        m.reserve(2048);
        for (std::uint16_t i = 0; i < 2048; ++i) {
            m.insert(QString::fromUtf8(kBip39En[i]), i);
        }
        return m;
    }();
    return kIndex;
}

}  // namespace

QString seed_to_phrase(const Seed& seed) {
    // Build a 33-byte "entropy + checksum" buffer:
    // [seed(32)][SHA-256(seed)[0]] (top 8 bits of the digest = checksum)
    std::array<std::uint8_t, 32> digest{};
    crypto_hash_sha256(digest.data(), seed.data(), seed.size());

    std::array<std::uint8_t, 33> ent{};
    std::memcpy(ent.data(), seed.data(), 32);
    ent[32] = digest[0];

    // Walk 264 bits in 11-bit chunks (24 of them).
    QStringList words;
    words.reserve(24);
    for (int i = 0; i < 24; ++i) {
        std::uint16_t v = 0;
        for (int b = 0; b < 11; ++b) {
            const int bit  = i * 11 + b;
            const int byte = bit >> 3;
            const int off  = 7 - (bit & 7);
            v = static_cast<std::uint16_t>((v << 1) | ((ent[byte] >> off) & 1));
        }
        words.append(QString::fromUtf8(kBip39En[v]));
    }
    return words.join(' ');
}

std::optional<Seed> phrase_to_seed(const QString& phrase) {
    QStringList tokens = phrase
        .normalized(QString::NormalizationForm_KD)
        .toLower()
        .trimmed()
        .split(QRegularExpression("\\s+"), Qt::SkipEmptyParts);
    if (tokens.size() != 24) return std::nullopt;

    const auto& idx = word_index();
    std::array<std::uint8_t, 33> ent{};
    int bitpos = 0;
    for (const QString& w : tokens) {
        const auto it = idx.constFind(w);
        if (it == idx.constEnd()) return std::nullopt;
        std::uint16_t v = *it;
        for (int b = 10; b >= 0; --b) {
            const int byte = bitpos >> 3;
            const int off  = 7 - (bitpos & 7);
            ent[byte] = static_cast<std::uint8_t>(
                ent[byte] | (((v >> b) & 1) << off));
            bitpos++;
        }
    }
    Seed seed{};
    std::memcpy(seed.data(), ent.data(), 32);

    std::array<std::uint8_t, 32> digest{};
    crypto_hash_sha256(digest.data(), seed.data(), seed.size());
    if (digest[0] != ent[32]) return std::nullopt;   // checksum mismatch
    return seed;
}

}  // namespace fb::desktop

File: client-web/ui/bip39.js
// SPDX-License-Identifier: AGPL-3.0-or-later
//
// Minimal BIP39 encode/decode for the 32-byte FinBit identity seed.
//
//   32-byte seed (256 bits) + SHA-256(seed)[0..7] (8 checksum bits) = 264 bits
//                                                                  = 24 × 11-bit chunks
//   Each 11-bit chunk indexes into the 2048-word English wordlist.
//
// Round-trip: seed → 24 words → seed.  Recovery from 24 words is rejected
// if the embedded checksum doesn't match — protects against typos.
//
// We accept the BIP39 canonical wordlist and produce 24-word phrases.
// Smaller word counts (12 / 18) aren't supported; we always have 32 bytes.
//
// SHA-256 comes from WebCrypto (browsers + Node 22).

import { BIP39_EN } from "./bip39_wordlist.js";

if (BIP39_EN.length !== 2048) {
    throw new Error(`bip39: wordlist must be exactly 2048 entries (got ${BIP39_EN.length})`);
}

// Build a fast lookup: word → index. Throws if a word is unknown.
const _index = new Map(BIP39_EN.map((w, i) => [w, i]));
function indexOf(w) {
    const i = _index.get(w);
    if (i === undefined) throw new Error(`bip39: unknown word "${w}"`);
    return i;
}

async function sha256(bytes) {
    return new Uint8Array(await crypto.subtle.digest("SHA-256", bytes));
}

// 32-byte seed → 24-word phrase (space-separated, lowercase).
export async function seedToPhrase(seed) {
    if (!(seed instanceof Uint8Array) || seed.byteLength !== 32) {
        throw new Error("seedToPhrase: seed must be Uint8Array(32)");
    }
    // 256 bits + 8 checksum bits = 264 bits = 33 bytes laid out as
    // [seed(32)][checksum_top_8_bits_of_sha256_of_seed (1 byte)].
    const cs = (await sha256(seed))[0];
    const ent = new Uint8Array(33);
    ent.set(seed, 0);
    ent[32] = cs;

    // Walk 264 bits in 11-bit chunks (24 of them).
    const words = new Array(24);
    for (let i = 0; i < 24; i++) {
        let v = 0;
        for (let b = 0; b < 11; b++) {
            const bit = i * 11 + b;          // bit position in `ent`
            const byte = ent[bit >> 3];
            const off  = 7 - (bit & 7);
            v = (v << 1) | ((byte >> off) & 1);
        }
        words[i] = BIP39_EN[v];
    }
    return words.join(" ");
}

// 24-word phrase → 32-byte seed. Throws on:
//   * not exactly 24 whitespace-separated tokens
//   * unknown word
//   * checksum mismatch
//
// Whitespace and case are normalized; capitalization is ignored.
export async function phraseToSeed(phrase) {
    const words = phrase
        .normalize("NFKD")
        .toLowerCase()
        .trim()
        .split(/\s+/)
        .filter(Boolean);
    if (words.length !== 24) {
        throw new Error(`bip39: phrase must have 24 words (got ${words.length})`);
    }
    // Pack 24 × 11-bit indices into a 264-bit big-endian buffer.
    const ent = new Uint8Array(33);
    let bitpos = 0;
    for (const w of words) {
        let v = indexOf(w);
        for (let b = 10; b >= 0; b--) {
            const byte = bitpos >> 3;
            const off  = 7 - (bitpos & 7);
            ent[byte] |= ((v >> b) & 1) << off;
            bitpos++;
        }
    }
    const seed = ent.slice(0, 32);
    const expected = (await sha256(seed))[0];
    if (ent[32] !== expected) {
        throw new Error("bip39: checksum mismatch — typo in the recovery phrase?");
    }
    return seed;
}
